python/tk_multi_workfiles/entity_models/deferred_model.py

```diff
@@ -0,0 +1,588 @@
+          a regular Shotgun Entity Model.
+        - Secondary requests are deferred until the point the data really needs
+          to be fetched, typically when a leaf in the primary model is expanded.
+          The primary tree is then extended with child items dynamically retrieved
+          on demand with the deferred queries.
+
+        Deferred queries need to specify:
+          - The target entity type for the query, e.g. 'Task'.
+          - The field name to link the secondary query to the primary one, e.g.
+            'entity'.
+        A sub-hierarchy can be defined with a list of fields, e.g. ['step'].
+        If needed, additional filters can be specified for deferred queries.
+
+        :param entity_type: The type of the entities that should be loaded into this model.
+        :param filters: A list of filters to be applied to entities in the model - these
+                        will be passed to the Shotgun API find() call when populating the
+                        model
+        :param hierarchy: List of Shotgun fields that will be used to define the structure
+                          of the items in the model.
+        :param fields: List of Shotgun fields to populate the items in the model with.
+                       These will be passed to the Shotgun API find() call when populating
+                       the model.
+        :param deferred_query: A dictionary with the `entity_type`, `link_field`
+                               `filters` and `hierarchy` keys, allowing to run a
+                               Shotgun sub-query for a given entity in this model.
+        """
+        # Basic sanity check that we do have a deferred query, so we don't have
+        # to test it everywhere.
+        if not deferred_query:
+            raise ValueError("A non empty deferred query is required")
+
+        self._deferred_query = deferred_query
+        self._task_step_icons = {}
+        super(ShotgunDeferredEntityModel, self).__init__(
+            entity_type,
+            filters,
+            hierarchy,
+            fields,
+            *args,
+            **kwargs
+        )
+        self._deferred_cache = ShotgunDataHandlerCache()
+        self._deferred_models = {}
+
+    @property
+    def deferred_query(self):
+        """
+        :returns: The deferred query for this model, if any.
+        """
+        return self._deferred_query
+
+    @property
+    def represents_tasks(self):
+        """
+        :returns: True if this model represents Tasks through deferred query.
+        """
+        return self._deferred_query["entity_type"] == "Task"
+
+    @property
+    def supports_step_filtering(self):
+        """
+        :returns: True if Step filtering can be used with this model
+        """
+        # If don't have steps in the fields we query from Shotgun, we assume
+        # step filtering should be disabled.
+        return "step" in self._deferred_query["hierarchy"]
```

Comment:
Given that most of this code is designed for Tasks, I think it will be overkill to support other cases here. Or it would mean a lot of changes to support a general case.